### PR TITLE
SMES improvements

### PR DESCRIPTION
**Pull request by Tehsapper**
Adds a new "automatic" charging mode for SMES units, which lets them charge using any amount of available power that is less or equal to their input level. The reason for that change is that SMESes were actually incredibly bad at storing volatile energy (such as solar energy), because they would only charge at set input level if there was more excess power than that value, which practically meant wasting a lot of generated power. This led to an interesting situation where any competent engineer would rather wire the solars directly to the power grid, as APCs were better at storing energy efficiently, leaving all solar SMES units on all maps virtually useless.
This change allows solar SMES to be actually useful, saving almost all generated solar power for future use.

The old ("manual") mode is still there, because it might be useful for fine-tuning SMES power collection.

Tweaked upgraded SMES capacity and input/output levels to make more of a difference.
SMES UI now displays a message if input terminal or output power cable are missing.

Fixes a millenia old bug where you could deconstruct/construct SMES for free "roundstart" power: newly constructed SMES start out with zero charge.
Fixes SMES not connecting to output powernet upon construction. Also fixes SMES being able to output (and lose) charge even if there is no powernet.
Fixes some typos related to SMES.

:cl:
 * rscadd: added automatic charge mode for SMES, allowing them to charge using any amount of available power that is less or equal to their input level. The old ("manual") mode is still there.
 * tweak: upgraded SMES have better capacity and max input/output.
 * tweak: SMES UI now shows if input terminal or output power cable are missing.
 * bugfix: fixed being able to reconstruct SMES for free power. New SMES start out with zero charge (roundstart SMES are not affected).
 * bugfix: fixed SMES not connecting to output powernet upon construction.
 * bugfix: fixed SMES being able to output (and lose) charge even if there is no powernet.

### CODE REVIEW

**Comment by Aurx** [comment]
The really competent engineers just wired the main engine to the grid and short-circuited the solar SMESes, but this is a step forwards for people who don't like engie-fried assistant.

**Comment by CptWad** [comment]
does this mean if I output 70 million it charges the entire SMES in seconds or is there an upper limit?

**Comment by Tehsapper** [comment]
Input level (the one you had to set for SMES to charge) is the upper limit, which in turn is limited by its maximum value (200k for an unupgraded SMES; 600k if fully upgraded)

**Comment by Blithering** [comment]
Does this apply to portable SMESes like bagel uses?

**Comment by ghost** [comment]
Saving a life.

**Comment by Tehsapper** [comment]
The auto charging should work, though free power abuse bugfix does not apply to them.

File: code/modules/power/battery.dm [hunk 1]
```diff
@@ -192,12 +197,15 @@ var/global/list/battery_online =	list(
...
+	data["chargeLoad"] = round(loadcharge)
 	data["chargeLevel"] = chargelevel
-	data["chargeMax"] = smes_input_max
+	data["chargeMax"] = max_input
 	data["outputOnline"] = online
 	data["outputLevel"] = output
-	data["outputMax"] = smes_output_max
+	data["outputMax"] = max_output
 	data["outputLoad"] = round(loaddemand)
+	data["hasInput"] = terminal ? 1 : 0;
```
Comment (DamianX) [inline code/modules/power/battery.dm]: This won't work for portable SMESs, the terminal is owned by a different machine

**Review by DamianX — CHANGES_REQUESTED** [review]

File: code/modules/power/battery.dm (hunk 1 above)
Comment (Tehsapper) [reply code/modules/power/battery.dm]: A bit busy IRL, will look into in a couple of days.

**Comment by Chronakai** [comment]
So in essence you could just Set It and Forget It at 200k roundstart and flip Output to whatever? Why even have an Input setting anyway?

**Comment by Aurx** [comment]
Assuming that SMESes can draw power before other things, setting the draw to max could lead to the SMES taking power that something else should have had.  Especially if you short-circuit the SMES.

File: code/modules/power/battery.dm [hunk 2]
```diff
@@ -51,16 +55,17 @@ var/global/list/battery_online =	list(
 
 	var/output = 50000
 	var/lastout = 0
+	var/loadcharge = 0
 	var/loaddemand = 0
 	var/capacity = 5e6 //Max stored charge
 	var/charge = 1e6 //Stored charge
-	var/charging = 0 //Are we currently taking charge in?
+	var/charging = BATTERY_NO_CHARGE //Are we currently taking charge in?
```
Comment (PJB3005) [inline code/modules/power/battery.dm]: `BATTERY_NO_CHARGE` is for `chargemode`. Use `TRUE`/`FALSE` here.

File: nano/templates/smes.tmpl [hunk 3]
```diff
@@ -40,8 +43,27 @@
...
+		{{if data.chargeLoad > 0}}
+			<span class='good'>+{{:data.chargeLoad}} W</span>
+		{{else}}
+			<span class='average'>0 W</span>
+		{{/if}}
+	</div>
+{{else}}
+<div class="item">
+	<div class="itemContent">
+		<span class='bad'>No input terminal detected!</span>
```
Comment (PJB3005) [inline nano/templates/smes.tmpl]: Use warning backgrounds for these like on the APC GUI.

**Review by PJB3005 — DISMISSED** [review]

**Comment by DamianX** [comment]
I hereby approve of my own changes done to this PR

Someone pls review and merge

**Review by DamianX — APPROVED** [review]

**Comment by Probe1** [comment]
This is a really good PR!  Why did it sit for a month..

File: code/modules/power/battery.dm [hunk 4]
```diff
@@ -119,15 +124,15 @@ var/global/list/battery_online =	list(
...
+			if ((chargemode == BATTERY_MANUAL_CHARGE && excess >= chargelevel) || (chargemode == BATTERY_AUTO_CHARGE && excess > 0))
 				chargecount++
 			else
 				chargecount = 0
 		else
 			chargecount = 0
 
 	// Output
-	if (online)
+	if (online && get_powernet()) // how can discharge be real if our powernet isn't real
```
Comment (Probe1) [inline code/modules/power/battery.dm:135]: 😁

File: code/modules/power/battery.dm [hunk 5]
```diff
@@ -232,9 +240,14 @@ var/global/list/battery_online =	list(
 		return 0 // Do not update ui
 
 	if( href_list["cmode"] )
-		chargemode = !chargemode
-		if(!chargemode)
-			charging = 0
+		switch( href_list["cmode"])
```
Comment (Probe1) [inline code/modules/power/battery.dm:243]: That fucking space

**Review by Probe1 — APPROVED** [review]
Looks good, playtesting looked good though it may have some funny issues with overcharging stationside SMES by just spamming auto max max and ignoring the internal engineering SMES.

**Pull request merged**